File: Competitive_Coding/fib.c
#include<stdio.h>
#define NIL -1

int table[10];

void initialise(){
  int i;
  for ( i = 0; i < 10; i++) {
    table[i]=NIL;
  }
}

int fib(int n){
  if(n==0 || n==1)
  return (n);
  else{
  if(table[n-1]==NIL)
  table[n-1]=fib(n-1);

  if(table[n-2]==NIL)
  table[n-2]=fib(n-2);

  table[n]=table[n-1]+table[n-2];

  return(table[n]);
}
}



int main(){
  int n=7;
  initialise();
  printf("fib(7) :%d",fib(n));
  return 0;
}

File: Algorithm_Java/Dynamic_programming/Fibonacci/Fib.java
package Fibonacci;

import java.util.Map;

public class Fib {

    int fibRecursion(int n){
        if(n==0||n==1)
        return 1;
        else
        return fibRecursion(n-1)+fibRecursion(n-2);

    }
    //top-down 
    int fibMemoization(int n,Map<Integer,Integer> table){
      
        if(!table.containsKey(n))
            table.put(n, fibMemoization(n-1, table)+fibMemoization(n-2, table));
        return table.get(n);
    }

    int fibTabulation(int n,Map<Integer,Integer> table){
        
        for(int i=2;i<=n;i++){
            table.put(i, table.get(i-1)+table.get(i-2));
        }
        return table.get(n);
    }
    
}
